src/backend/distributed/commands/alter_table.c

```diff
@@ -1572,3 +1580,95 @@ ExecuteQueryViaSPI(char *query, int SPIOK)
 		ereport(ERROR, (errmsg("could not finish SPI connection")));
 	}
 }
+
+
+/*
+ * SwitchToSequentialAndLocalExecutionIfRelationNameTooLong generates the longest shard name
+ * on the shards of a distributed table, and if exceeds the limit switches to sequential and
+ * local execution to prevent self-deadlocks.
+ *
+ * In case of a RENAME, the relation name parameter should store the new table name, so
+ * that the function can generate shard names of the renamed relations
+ */
+void
+SwitchToSequentialAndLocalExecutionIfRelationNameTooLong(Oid relationId,
+														 char *finalRelationName)
+{
+	if (!IsCitusTable(relationId))
+	{
+		return;
+	}
+
+	char *longestShardName = GetLongestShardName(relationId, finalRelationName);
```

Comment:
This fails for zero shard tables, we should fix it:
```SQL
create table test (a int);
select create_distributed_table('test', 'a', 'append');
 alter table test rename to aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111l;
NOTICE:  identifier "aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111l" will be truncated to "aaaaaa111111111aaaaaa111111111aaaaaa111111111aaaaaa111111111aaa"
ERROR:  could not find valid entry for shard 0
```

Check `SwitchToSequentialAndLocalExecutionIfIndexNameTooLong` for an example, and please add regression tests for both rename and create index. Thanks